doc/search: Patch search to display results bigger if hitting return and work on mobil

kfessel opened the pull request:
### Contribution description

make the search work on Mobil #7494 

make search not reload on hitting return, but display the results in content-part of the window

### Testing procedure

I used Firefox inspector to test

### Issues/PRs references

Fixes #14409 doc/search: hitting return while searching on api.riot-os.org reloads current page #14409
Fixes #7494

This is in conflict with  #14411 only one should be applied, I prefer this one

kfessel commented:
I would like to add @maribu to the reviewers of this PR.

maribu commented:
Doesn't work on my machine (with latest stable version of Doxygen). But it does prevent the page reload, which would safe me 10 minutes a day of retyping my search string. I just keep reflexively hitting the return key after I entered the search string the first time.

I would already be more than happy with the new behavior of ignoring the return key, but that could be already achieved with `onsubmit="return false;"`.

maribu commented:
How about this: We merge https://github.com/RIOT-OS/RIOT/pull/14411 swiftly (as it is simple and solves the more pressing issue reliable). This PR can be rebased on top and merged in addition, once it works reliable?

maribu commented:
I can confirm that this works if I "patch" the live website using Firefox's Inspector (or more: patch the HTML in the buffer of my local Firefox, not the HTML actually hosted on the server).

kfessel commented:
https://kfessel.github.io/RIOT/html/  << there is this, and also a patch (read workaround) for #7494

kfessel commented:
this is currently conflicting since #14411 got applyed which containt the patial solution to the problem

maribu commented:
> this is currently conflicting since #14411 got applyed which containt the patial solution to the problem

You can fix this by rebasing on top.

```
git checkout master
git pull upstream master
git checkout patch-doxygen-search2
git rebase master
# This will show a merge conflict. You can just take the file from your old commit in this case
git checkout ef08db51290a1d36d4d75284ee6030a84147b01d -- doc/doxygen/header.html
git add doc/doxygen/header.html
git rebase --continue
git push --force-with-lease
```

miri64 commented:
> https://kfessel.github.io/RIOT/html/ << there is this, and also a patch (read workaround) for #7494

See also https://github.com/RIOT-OS/RIOT/issues/7494#issuecomment-652954600

kfessel commented:
i will rebase and include the mobil patch

kfessel commented:
i rebased on master and added the patch for #7494

view on https://kfessel.github.io/RIOT/html/

If you like to have things moved around or resized on mobil (20px down) and result.width = screen.width please tell me

miri64 commented:
Yepp, looking good!

![Screenshot_20200702-162231](https://user-images.githubusercontent.com/675644/86371103-cbd82f80-bc80-11ea-9679-f1fa6d436797.png)

I also took the liberty to adapt OP for automatic closure of the referenced issues.

kfessel commented:
there is an issue with the display search results big: they point to ".. " i just copy them from the iframe which has an other base url so its the wrong dir

kfessel commented:
**do not merge** (for now)

miri64 commented:
This should care of that :-).

kfessel commented:
i am pretty sure it works as intended now please check

maribu commented:
Maybe you could hide the search results windows on mobile devices (which there remains unreadable / empty anyway). This should do the trick:

``` diff
diff --git a/doc/doxygen/src/css/riot.css b/doc/doxygen/src/css/riot.css
index 85e65fde97..a3f9219f02 100644
--- a/doc/doxygen/src/css/riot.css
+++ b/doc/doxygen/src/css/riot.css
@@ -519,3 +519,8 @@ dl.todo > dt {
 table.retval > tbody > tr > td.paramname {
   padding-right: 2em;
 }
+@media screen and (max-width: 767px) {
+  #MSearchResultsWindow {
+    display: none;
+  }
+}
diff --git a/doc/doxygen/src/css/riot.less b/doc/doxygen/src/css/riot.less
index 6ce63d24f2..ba8202dc6b 100644
--- a/doc/doxygen/src/css/riot.less
+++ b/doc/doxygen/src/css/riot.less
@@ -461,3 +461,8 @@ dl.todo > dt {
 table.retval > tbody > tr > td.paramname {
   padding-right: 2em;
 }
+@media screen and (max-width: 767px) {
+  #MSearchResultsWindow {
+    display: none;
+  }
+}
```

Of course, the results will still show up on mobile when confirming the search; only the live results will not show (which is broken anyway on mobile).

kfessel commented:
its not empty but on some screen sizes, it appears under the keyboard. (and there is a small issue that it appears delayed (internals of doxygen search.js) and therefor it is moved and resized at the keyupevent after it is displayed.

maribu commented:
> And there is a small hickup that it appears delayed

Indeed. Sorry for the noise. Maybe it is time to buy a faster smartphone...

kfessel commented:
no a faster Phone would not the delay is by design in the search.js by doxygen, when u press a key (keyup); if there is a timeout it is deleted and at the and of the funktion it creates a new one that will search and show the result after about 500ms (if you do not press a key (otherwise timeout delete and a new one)

but the search (after 500ms) also is the part that makes the result window showup (and possition the bad way doxygen likes it), and only after it is shown i can move it to the place i and you like.

maribu commented on doc/doxygen/header.html:
```diff
@@ -30,6 +30,15 @@
     <![endif]-->
+        <script>
+            function displaySR(){
+            var rhtml=(document.getElementById('MSearchResults').contentWindow.document.body.innerHTML).replace(/href=\"\.\.\//g,'href="');
```
This results in a Content Security Policy violation on Firefox when accessing a locally generated documentation.

The origin of this script is `file:///`, the search results are an `iframe` with `javascript:void(0)` as origin. Apparently, this is considered cross origin.

maribu replied:
Note: Just moving the body of the function directly into the `onsubmit=""` doesn't help. (Which makes sense, it still has origin `file:///`.)

kfessel replied:
OK, for me this seems like a fix is not worth it

kfessel replied:
i like the method of extracting the result from the Iframe into a div

kfessel commented:
i found a place to put my functions (the searchBox objekt)

maribu commented on doc/doxygen/header.html:
```diff
@@ -32,12 +32,25 @@
             function displaySR(){
+            searchBox.Search();
             var rhtml=(document.getElementById('MSearchResults').contentWindow.document.body.innerHTML).replace(/href=\"\.\.\//g,'href="');
             document.getElementById('MSearchResultsWindow').style.display='none';
```
Now it fails here :-/ Seems like the JavaScript has bit rotten a bit. Anyway, that is not the concern of this PR. I'd say this can go in. It should still work on the live system, which is more important.

maribu commented:
@miri64: Can this go in despite feature freeze?

miri64 commented:
> @miri64: Can this go in despite feature freeze?

Hard feature freeze is next week ;-) and for soft feature freeze; I don't see this as a high impact change... do you oO?

miri64 commented:
One of the commit messages is too long. See Murdock log.

miri64 commented:
@kfessel please use `git rebase -i master` to `reword` https://github.com/RIOT-OS/RIOT/pull/14412/commits/f600d2eca5c34ea4744ea6ed71b2410c5848f3cb with a summary line that is <72 characters. You may use more detailed description in the body of the commit message. Here is an example

```
doc: search improvements

- display results bigger if hitting return
- make it work on mobile
```

kfessel commented:
Are you with me some commits and making this 2 or three?

miri64 commented:
Yes, it is welcomed even :-). I guess one for the results to show bigger and one for the search box on mobile would be good.

kfessel commented:
done

miri64 commented:
There is a typo in your commit message: mobil*e*

kfessel commented:
fixed the typo

miri64 commented:
Then let's go!
